`TestCLITestSuite/TestCmdQueryTally` failed test-race

test-race failed on 

```
WARNING: DATA RACE
Read at 0x00c002c67c20 by goroutine 992:
  cosmossdk.io/store/rootmulti.(*Store).LastCommitID()
      /home/runner/go/pkg/mod/cosmossdk.io/store@v1.1.2/rootmulti/store.go:448 +0x49
  cosmossdk.io/store/rootmulti.(*Store).LatestVersion()
      /home/runner/go/pkg/mod/cosmossdk.io/store@v1.1.2/rootmulti/store.go:443 +0x26
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).CreateQueryContext()
      /home/runner/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v0.52.1/baseapp/abci.go:1267 +0x1a9
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func1()
      /home/runner/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v0.52.1/baseapp/grpcserver.go:49 +0x1aa
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func2.ChainUnaryServer.2.1()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/chain.go:48 +0xcd
  github.com/grpc-ecosystem/go-grpc-middleware/recovery.UnaryServerInterceptor.func1()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/recovery/interceptors.go:33 +0x164
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func2.ChainUnaryServer.2()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/chain.go:53 +0x222
  github.com/celestiaorg/celestia-app/v7/x/signal/types._Query_VersionTally_Handler()
      /home/runner/work/celestia-app/celestia-app/x/signal/types/query.pb.go:460 +0x1e6
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func2()
      /home/runner/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v0.52.1/baseapp/grpcserver.go:81 +0x1ad
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.79.1/server.go:1426 +0x1a69
  google.golang.org/grpc.(*Server).handleStream()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.79.1/server.go:1830 +0x1772
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.79.1/server.go:1061 +0x149


==================
    testing.go:1617: race detected during execution of test
=== NAME  TestCLITestSuite
    network.go:97: tearing down testnode
--- FAIL: TestCLITestSuite (4.49s)
    --- PASS: TestCLITestSuite/TestCmdGetUpgrade (0.01s)
    --- FAIL: TestCLITestSuite/TestCmdQueryTally (0.03s)
FAIL
FAIL	github.com/celestiaorg/celestia-app/v7/x/signal/cli	5.380s
```

https://github.com/celestiaorg/celestia-app/actions/runs/22228726555/job/64302224284?pr=6611

--- a/baseapp/abci.go
+++ b/baseapp/abci.go
@@ -961,7 +961,17 @@ func (app *BaseApp) Commit() (*abci.ResponseCommit, error) {
 		rms.SetCommitHeader(header)
 	}
 
+	// Hold the write lock across both the store commit and the check state
+	// reset. This prevents data races with concurrent gRPC queries that read
+	// LatestVersion() and checkState in CreateQueryContext(). The store's
+	// cms.Commit() writes lastCommitInfo which is read by LatestVersion().
+	//
+	// NOTE: This is safe because CometBFT holds a lock on the mempool for
+	// Commit. Use the header from this latest block.
+	app.checkStateMu.Lock()
 	app.cms.Commit()
+	app.setState(execModeCheck, header)
+	app.checkStateMu.Unlock()
 
 	resp := &abci.ResponseCommit{
 		RetainHeight: retainHeight,
@@ -979,14 +989,6 @@ func (app *BaseApp) Commit() (*abci.ResponseCommit, error) {
 			}
 		}
 	}
-
-	// Reset the CheckTx state to the latest committed.
-	//
-	// NOTE: This is safe because CometBFT holds a lock on the mempool for
-	// Commit. Use the header from this latest block.
-	app.checkStateMu.Lock()
-	app.setState(execModeCheck, header)
-	app.checkStateMu.Unlock()
 
 	app.finalizeBlockState = nil
 
@@ -1264,7 +1266,15 @@ func (app *BaseApp) CreateQueryContext(height int64, prove bool) (sdk.Context, e
 		qms = app.cms.(storetypes.MultiStore)
 	}
 
+	// Hold checkStateMu.RLock across both the LatestVersion() read and the
+	// checkState header read. Commit() updates both the store's lastCommitInfo
+	// (read by LatestVersion) and checkState under the corresponding write lock,
+	// so this prevents data races with concurrent block commits.
+	app.checkStateMu.RLock()
 	lastBlockHeight := qms.LatestVersion()
+	header := app.checkState.Context().BlockHeader()
+	app.checkStateMu.RUnlock()
+
 	if lastBlockHeight == 0 {
 		return sdk.Context{}, errorsmod.Wrapf(sdkerrors.ErrInvalidHeight, "%s is not ready; please wait for first block", app.Name())
 	}
@@ -1298,11 +1308,6 @@ func (app *BaseApp) CreateQueryContext(height int64, prove bool) (sdk.Context, e
 				"failed to load state at height %d; %s (latest height: %d)", height, err, lastBlockHeight,
 			)
 	}
-
-	// branch the commit multi-store for safety
-	app.checkStateMu.RLock()
-	header := app.checkState.Context().BlockHeader()
-	app.checkStateMu.RUnlock()
 	ctx := sdk.NewContext(cacheMS, header, true, app.logger).
 		WithMinGasPrices(app.minGasPrices).
 		WithGasMeter(storetypes.NewGasMeter(app.queryGasLimit)).

